In [1]:
import pandas as pd

In [68]:
brands_url = "https://drive.google.com/file/d/1m1ThDDIYRTTii-rqM5SEQjJ8McidJskD/view?usp=share_link"
brands_path = "https://drive.google.com/uc?export=download&id="+brands_url.split("/")[-2]
brands = pd.read_csv(brands_path)

orderlines_url = "https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing" # orderlines.csv
orderlines_path = "https://drive.google.com/uc?export=download&id="+orderlines_url.split("/")[-2]
orderlines = pd.read_csv(orderlines_path)

orders_url = "https://drive.google.com/file/d/1Vu0q91qZw6lqhIqbjoXYvYAQTmVHh6uZ/view?usp=share_link"
orders_path = "https://drive.google.com/uc?export=download&id="+orders_url.split("/")[-2]
orders = pd.read_csv(orders_path)

products_url = "https://drive.google.com/file/d/1afxwDXfl-7cQ_qLwyDitfcCx3u7WMvkU/view?usp=share_link"
products_path = "https://drive.google.com/uc?export=download&id="+products_url.split("/")[-2]
products = pd.read_csv(products_path)

In [21]:
products.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


In [37]:
#products duplicated values 
# False    10580
# True      8746
# dtype: int64
products.duplicated().value_counts()

#selecting only the duplicated values if the content of the same row repeats in all columns
#for now, thhere are some differences in some rows, so I will not drop them
# products[products.duplicated()]
# products[products['name'].duplicated()]

False    10580
True      8746
dtype: int64

#products.info()
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)

In [71]:
#changing data type of price and promo_price columns
# products['price'] = products['price'].astype(float)

# products['price'] = pd.to_numeric(products['price'])
# select rows to find values that have two dots in the price column

# products.loc[products.price.str.contains("\d+\.\d+\.\d+"), :]



# products.price.str.contains("\d+\.\d+\.\d+").value_counts() # false 18849, true 431



# products['price'] = products.loc[~products.price.isna(), :]
# orders.loc[~orders.total_paid.isna(), :]







# two_dot_numbers = products.loc[products.price.str.contains("\d+\.\d+\.\d+")]
# two_dot_order = orderlines.loc[orderlines.unit_price.str.contains("\d+\.\d+\.\d+"), "id_order"]

Series([], dtype: int64)

In [91]:
import re
#select values that have NA/NAN values in the price column to observe them
# 46 NANs in the price column
products.loc[products.price.isna(), :]

# replaced NANs with 0
products['price'] = products['price'].fillna(0)
len(products.loc[products.price.isna(), :]) # 0 NANs in the price column

# products['price'] = pd.to_numeric(products['price']) # ValueError: Unable to parse string "1.639.792"
# select rows to find values that have two dots in the price column
# products.loc[products['price'].contains("\d+\.\d+\.\d+"), :]

# two_dot_numbers = products.loc[products.price.str.contains("\d+\.\d+\.\d+")]
replace_dots = products['price'].replace(to_replace = "\.", value = "", regex = True)
products['price'] = replace_dots
products['price'].head(20)





0     5999
1       59
2       59
3       25
4     3499
5       45
6     1899
7     3699
8       74
9       35
10      35
11    4999
12      35
13      89
14      20
15    2495
16      89
17    6999
18     999
19      35
Name: price, dtype: object

percentage of missing values 
"price": 0.00238%
'type: 0.002587%

In [93]:
products.head(30)

sku                                               name  \
0     RAI0007                  Silver Rain Design mStand Support   
1     APP0023                  Apple Mac Keyboard Keypad Spanish   
2     APP0025                   Mighty Mouse Apple Mouse for Mac   
3     APP0072      Apple Dock to USB Cable iPhone and iPod white   
4     KIN0007        Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
5     APP0073     Apple Composite AV Cable iPhone and iPod white   
6     KIN0008        Mac Memory Kingston 1GB 667MHz DDR2 SO-DIMM   
7     KIN0009        Mac Memory Kingston 2GB 800MHz DDR2 SO-DIMM   
8   KIN0001-2  Mac memory Kingston 4GB (2x2GB) 667MHz DDR2 SO...   
9     APP0100             Apple Adapter Mini Display Port to VGA   
10    APP0101              Apple Mini DisplayPort to DVI Adapter   
11    SEN0021  Sennheiser CX 300-II Precision headphones iPho...   
12    APP0111                     Apple USB Ethernet Adapter Mac   
13    APP0156              45W MagSafe charger Apple MacBook Air   
14    MOS0013         Adapter Moshi FireWire 400 to FireWire 800   
15    MOS0021               Clearguard Moshi MacBook Pro and Air   
16    APP0206  60W MagSafe charger Apple MacBook and MacBook ...   
17    HGD0007       Henge Docks Docking Station MacBook Pro 13 "   
18    MUV0009    Muvit Back Clear Case iPhone 4 / 4S Transparent   
19    APP0234                        Apple Dock Connector to VGA   
20    APP0247    85W MagSafe charger Apple MacBook Pro 15 "/ 17"   
21    APP0233             Apple iPad Camera Connection connector   
22    JMO0014                   Just Mobile Upstand iPad support   
23    MOB0007            Mobee Magic Charger Charger Magic Mouse   
24    APP0277         Apple Wireless Keyboard Keyboard (OEM) Mac   
25    MOP0011  Mophie Juice Pack Air Battery Case iPhone 4 / ...   
26    APP0280                      Apple HDMI to HDMI Cable 1.8m   
27    OND0069                           Adapter mini DVI to HDMI   
28    SNN0007              Sonnet Tempo SATA PCIe Mac and PC E2P   
29    KIN0025          Mac memory DIMM DDR3 2GB 1333MHz Kingston   

                                                 desc price promo_price  \
0        Aluminum support compatible with all MacBook  5999     499.899   
1           USB ultrathin keyboard Apple Mac Spanish.    59     589.996   
2                              mouse Apple USB cable.    59     569.898   
3               IPhone dock and USB Cable Apple iPod.    25     229.997   
4   2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  3499       31.99   
5   IPhone and iPod AV Cable Dock to Composite Video.    45     420.003   
6   1GB RAM Mac mini and iMac (2006/07) MacBook Pr...  1899     146.471   
7        2GB RAM iMac with Intel Core 2 Duo (Penryn).  3699     274.694   
8   RAM 4GB (2x2GB) Mac mini and iMac (2006/07) Ma...    74     669.904   
9   Adapter Mini Display Port to VGA MacBook and M...    35     330.003   
10  Adapter mini DisplayPort to DVI MacBook and Ma...    35     330.003   
11     Headphones iPhone iPad iPad 2 iPad 3 and iPod.  4999     449.878   
12                       USB to Ethernet adapter Mac.    35     330.003   
13     Charger MacBook Air MagSafe 45 W (MC747Z / A).    89     849.904   
14           FireWire 400 adapter cable FireWire 800.    20     179.891   
15  Keyboard Protector MacBook Pro 13-inch Retina ...  2495     249.889   
16  Charger MacBook and 13 inch MacBook MagSafe 60...    89     849.904   
17  Dock and support with multiple connections for...  6999     699.864   
18                  Case iPhone 4 / 4S polycarbonate.   999       7.986   
19                         Dock Connector to VGA IOS.    35     350.005   
20     Charger MacBook Pro MagSafe 85 W (MC556Z / A).    89     849.904   
21   IPad connector for digital cameras and SD cards.    35     349.896   
22        Support Retina iPad and iPad mini aluminum.  4995     399.905   
23                   Wireless Charger for Magic Mouse  4999     499.851   
24  Ultrathin keyboard App

sku            object
name           object
desc           object
price          object
promo_price    object
in_stock        int64
type           object
dtype: object

In [96]:
orders.head()
orders.groupby('state')['total_paid'].mean()

state
Cancelled          766.738125
Completed          341.381381
Pending            327.208899
Place Order        417.969775
Shopping Basket    729.252943
Name: total_paid, dtype: float64